Add a new module: check_configuration

carlocas opened the pull request:
|Related issue|
|---|
|[3731](https://github.com/wazuh/wazuh/issues/3731)|

## Description

The purpose of this module is to read the configuration files and check if has an error in them. Also, a feature if this module, is that it is independent of wazuh, so we can check the files even if wazuh isn't running.

## Configuration options

The module `check_configuration` has only two options. One of them is optional. This option `-f` is used to indicated the path to the file to validate. The other option is `-t` that indicates the type of file to validate, `manager`, `agent` and `remote`.

The default path for each option is `/var/ossec/etc/ossec.conf` for `manager` and `agent`, and `/var/ossec/etc/shared/default/agent.conf` for  `remoted`.

An example of the use of the module by CLI `./check_configuration -t manager`. Whit this command we validate the configuration file of the manager in `/var/ossec/etc/ossec.conf` so, the previous command will validate the same file as, `check_configuration -t manager -f /var/ossec/etc/ossec.conf`

## Logs/Alerts example

The input format of the JSON for the socket is the following:
```JSON
{
  "operation": "GET",
  "type": "request",
  "version": "3.10",
  "component": "check_configuration",
  "data":
    {
      "type": "manager",
      "file": "/var/ossec/etc/ossec.conf"
    },
}
```
As in the CLI, we have an optional field, in this case, `data.file`. If we don't fill this field the validator will check the file by default. The answer to the previusly input will be:
```JSON
{
  "error":"1",
  "data":[
    {
      "type":"ERROR",
      "message":"Invalid elements in node fields: NODE_IP."
    },
    {
      "type":"ERROR",
      "message":"Configuration error at '/var/ossec/etc/ossec.conf'."
    },
    {
      "type":"ERROR",
      "message":"Invalid configuration in Analysisd"
     }
    ]
}
```
## Tests

<!-- Minimum checks required -->
- Compilation without warnings in every supported platform
  - [x] Linux
- [x] Source installation
- [ ] Package installation
- [x] Source upgrade
- [ ] Package upgrade
- [x] QA templates contemplate the added capabilities

<!-- Depending on the affected OS -->
- Memory tests for Linux
  - [x] Scan-build report
  - [ ] Coverity
  - [x] Valgrind (memcheck and descriptor leaks check)
Almost all the leaks that appeared in this PR were related with multiples modules of configuration repeated in the file. These memory leaks were reported in this [PR](https://github.com/wazuh/wazuh/issues/3933). 
 ```
==19170== HEAP SUMMARY:
==19170==     in use at exit: 0 bytes in 0 blocks
==19170==   total heap usage: 1 allocs, 1 frees, 1,024 bytes allocated
==19170== 
==19170== All heap blocks were freed -- no leaks are possible
```
  - [x] AddressSanitizer
- Memory tests for macOS
  - [ ] Scan-build report
  - [ ] Leaks
  - [ ] AddressSanitizer

- Validate different types of configuration and check that doesn't write in `ossec.log` and doesn't have any memory leaks.
  - [x] Repeat modules of the configuration file.
We copy all the modules of the configuration file and introduces it, at the end of the file. We receive the following message from the validator:
```
Can't add more than one secure connection.
Configuration error at '/var/ossec/etc/ossec.conf'.
ERROR: Invalid configuration in Remoted
```
We remove the configuration module `remote` to get a successful validation. Once removed, we obtained a successful validation of the file.
  - [x] Introduces and element that isn't closed. 
We introduce the next invalid configuration and test, that, in every option receive an error.
```
<cluster>
    <name>wazuh</name>
    <node_name>node01</node_name>
    <node_type>master</node_type>
    <key></key>
    <port>1516</port>
    <bind_addr>0.0.0.0</bind_addr>
    <nodes>
        node>NODE_IP</node>
    </nodes>
    <hidden>no</hidden>
    <disabled>yes</disabled>
  </cluster>
```
We test the previous configuration in all the option avaliable `manager/agent/remote`. In all of them, we receive the next message. 
```
Error reading XML file '/var/ossec/etc/ossec.conf': XMLERR: Element 'nodes' not closed. (line 337).
```
Also, we don't detected any memory leaks, neither were logs write in `ossec.log`.
  - [x] Change disabled for enabled
```
 <cluster>
    <name>wazuh</name>
    <node_name>node01</node_name>
    <node_type>master</node_type>
    <key></key>
    <port>1516</port>
    <bind_addr>0.0.0.0</bind_addr>
    <nodes>
        <node>NODE_IP</node>
    </nodes>
    <hidden>no</hidden>
    <enabled>no</enabled>
  </cluster>
```
With this configuration we obtained an error. `enabled` is not a valid configuration. 
 
<!-- Checks for huge PRs that affect the product more generally -->
- [ ] Retrocompatibility with older Wazuh versions
- [x] Working on cluster environments
- [ ] Configuration on demand reports new parameters
- [ ] The data flow works as expected (agent-manager-api-app)
- [ ] Added unit tests (for new features)
- [ ] Stress test for affected components

carlocas commented:
While testing the fil for memory leaks, I discovered several memory leaks related to repeated configuration blocks. The report of Valgrind is the following:

```
==14688== HEAP SUMMARY:
==14688==     in use at exit: 9,025 bytes in 229 blocks
==14688==   total heap usage: 95,149 allocs, 94,920 frees, 60,177,325 bytes allocated
==14688== 
==14688== 6 bytes in 1 blocks are definitely lost in loss record 1 of 14
==14688==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==14688==    by 0x5B249B9: strdup (strdup.c:42)
==14688==    by 0x144C13: Read_Cluster (cluster-config.c:80)
==14688==    by 0x131F40: read_main_elements (config.c:175)
==14688==    by 0x132556: ReadConfig (config.c:274)
==14688==    by 0x132CD1: Test_Analysisd (config.c:486)
==14688==    by 0x112408: test_manager_conf (check_config.c:30)
==14688==    by 0x1120F6: main (check_configuration.c:116)
==14688== 
==14688== 7 bytes in 1 blocks are definitely lost in loss record 2 of 14
==14688==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==14688==    by 0x5B249B9: strdup (strdup.c:42)
==14688==    by 0x144E9A: Read_Cluster (cluster-config.c:100)
==14688==    by 0x131F40: read_main_elements (config.c:175)
==14688==    by 0x132556: ReadConfig (config.c:274)
==14688==    by 0x132CD1: Test_Analysisd (config.c:486)
==14688==    by 0x112408: test_manager_conf (check_config.c:30)
==14688==    by 0x1120F6: main (check_configuration.c:116)
==14688== 
==14688== 7 bytes in 1 blocks are definitely lost in loss record 3 of 14
==14688==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==14688==    by 0x5B249B9: strdup (strdup.c:42)
==14688==    by 0x14514D: Read_Cluster (cluster-config.c:120)
==14688==    by 0x131F40: read_main_elements (config.c:175)
==14688==    by 0x132556: ReadConfig (config.c:274)
==14688==    by 0x132CD1: Test_Analysisd (config.c:486)
==14688==    by 0x112408: test_manager_conf (check_config.c:30)
==14688==    by 0x1120F6: main (check_configuration.c:116)
==14688== 
==14688== 23 bytes in 1 blocks are definitely lost in loss record 4 of 14
==14688==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==14688==    by 0x5B249B9: strdup (strdup.c:42)
==14688==    by 0x13DA1E: Read_Global (global-config.c:721)
==14688==    by 0x1312F3: read_main_elements (config.c:78)
==14688==    by 0x132556: ReadConfig (config.c:274)
==14688==    by 0x14A5CD: Test_Maild (mail-config.c:25)
==14688==    by 0x1124B0: test_manager_conf (check_config.c:40)
==14688==    by 0x1120F6: main (check_configuration.c:116)
==14688== 
==14688== 30 bytes in 1 blocks are definitely lost in loss record 5 of 14
==14688==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==14688==    by 0x5B249B9: strdup (strdup.c:42)
==14688==    by 0x149EF8: Read_Authd (authd-config.c:235)
==14688==    by 0x131E72: read_main_elements (config.c:170)
==14688==    by 0x132556: ReadConfig (config.c:274)
==14688==    by 0x14A2AF: Test_Authd (authd-config.c:283)
==14688==    by 0x112380: test_manager_conf (check_config.c:22)
==14688==    by 0x1120F6: main (check_configuration.c:116)
==14688== 
==14688== 31 bytes in 1 blocks are definitely lost in loss record 6 of 14
==14688==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==14688==    by 0x5B249B9: strdup (strdup.c:42)
==14688==    by 0x149E72: Read_Authd (authd-config.c:232)
==14688==    by 0x131E72: read_main_elements (config.c:170)
==14688==    by 0x132556: ReadConfig (config.c:274)
==14688==    by 0x14A2AF: Test_Authd (authd-config.c:283)
==14688==    by 0x112380: test_manager_conf (check_config.c:22)
==14688==    by 0x1120F6: main (check_configuration.c:116)
==14688== 
==14688== 33 bytes in 1 blocks are possibly lost in loss record 7 of 14
==14688==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==14688==    by 0x5B249B9: strdup (strdup.c:42)
==14688==    by 0x14879A: Read_Rules (rules-config.c:411)
==14688==    by 0x13160F: read_main_elements (config.c:102)
==14688==    by 0x132556: ReadConfig (config.c:274)
==14688==    by 0x132CD1: Test_Analysisd (config.c:486)
==14688==    by 0x112408: test_manager_conf (check_config.c:30)
==14688==    by 0x1120F6: main (check_configuration.c:116)
==14688== 
==14688== 43 bytes in 1 blocks are definitely lost in loss record 9 of 14
==14688==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==14688==    by 0x5B249B9: strdup (strdup.c:42)
==14688==    by 0x125784: Read_Rootcheck (rootcheck-config.c:171)
==14688==    by 0x131748: read_main_elements (config.c:113)
==14688==    by 0x132556: ReadConfig (config.c:274)
==14688==    by 0x12744B: Test_Rootcheck (rootcheck-config.c:432)
==14688==    by 0x112492: test_manager_conf (check_config.c:38)
==14688==    by 0x1120F6: main (check_configuration.c:116)
==14688== 
==14688== 45 bytes in 1 blocks are definitely lost in loss record 10 of 14
==14688==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==14688==    by 0x5B249B9: strdup (strdup.c:42)
==14688==    by 0x12585C: Read_Rootcheck (rootcheck-config.c:174)
==14688==    by 0x131748: read_main_elements (config.c:113)
==14688==    by 0x132556: ReadConfig (config.c:274)
==14688==    by 0x12744B: Test_Rootcheck (rootcheck-config.c:432)
==14688==    by 0x112492: test_manager_conf (check_config.c:38)
==14688==    by 0x1120F6: main (check_configuration.c:116)
==14688== 
==14688== 51 bytes in 1 blocks are definitely lost in loss record 11 of 14
==14688==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==14688==    by 0x5B249B9: strdup (strdup.c:42)
==14688==    by 0x149BBB: Read_Authd (authd-config.c:210)
==14688==    by 0x131E72: read_main_elements (config.c:170)
==14688==    by 0x132556: ReadConfig (config.c:274)
==14688==    by 0x14A2AF: Test_Authd (authd-config.c:283)
==14688==    by 0x112380: test_manager_conf (check_config.c:22)
==14688==    by 0x1120F6: main (check_configuration.c:116)
==14688== 
==14688== 85 bytes in 3 blocks are definitely lost in loss record 12 of 14
==14688==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==14688==    by 0x5B249B9: strdup (strdup.c:42)
==14688==    by 0x146AB6: Read_Rules (rules-config.c:178)
==14688==    by 0x13160F: read_main_elements (config.c:102)
==14688==    by 0x132556: ReadConfig (config.c:274)
==14688==    by 0x132CD1: Test_Analysisd (config.c:486)
==14688==    by 0x112408: test_manager_conf (check_config.c:30)
==14688==    by 0x1120F6: main (check_configuration.c:116)
==14688== 
==14688== 4,092 bytes in 95 blocks are definitely lost in loss record 13 of 14
==14688==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==14688==    by 0x5B249B9: strdup (strdup.c:42)
==14688==    by 0x14827A: Read_Rules (rules-config.c:349)
==14688==    by 0x13160F: read_main_elements (config.c:102)
==14688==    by 0x132556: ReadConfig (config.c:274)
==14688==    by 0x132CD1: Test_Analysisd (config.c:486)
==14688==    by 0x112408: test_manager_conf (check_config.c:30)
==14688==    by 0x1120F6: main (check_configuration.c:116)
==14688== 
==14688== 4,531 bytes in 120 blocks are definitely lost in loss record 14 of 14
==14688==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==14688==    by 0x5B249B9: strdup (strdup.c:42)
==14688==    by 0x14879A: Read_Rules (rules-config.c:411)
==14688==    by 0x13160F: read_main_elements (config.c:102)
==14688==    by 0x132556: ReadConfig (config.c:274)
==14688==    by 0x132CD1: Test_Analysisd (config.c:486)
==14688==    by 0x112408: test_manager_conf (check_config.c:30)
==14688==    by 0x1120F6: main (check_configuration.c:116)
==14688== 
==14688== LEAK SUMMARY:
==14688==    definitely lost: 8,951 bytes in 227 blocks
==14688==    indirectly lost: 0 bytes in 0 blocks
==14688==      possibly lost: 33 bytes in 1 blocks
==14688==    still reachable: 41 bytes in 1 blocks
==14688==         suppressed: 0 bytes in 0 blocks
```

These memory leaks were caused because we have blocks of configuration repeated. This cause that, the variables were overwriting at the moment of reading the second module of the configuration. To prevent this problem, at the beginning of the function to read the configuration has been included a structure to check if it is the second module, and free the variables required.

Related commit:
[2e7e752](https://github.com/wazuh/wazuh/pull/4139/commits/2e7e7521af80b46051675210b60bbeaaa2564f95)

Molter73 commented on src/config/active-response.c:
```diff
@@ -257,14 +356,25 @@ int ReadActiveResponses(XML_NODE node, void *d1, void *d2)
+            mdebug1("Timeout is not allowed");
+            minfo(AR_NO_TIMEOUT, tmp_ar->ar_cmd->name);
+        } else {
+            snprintf(message, OS_FLSIZE + 1,
```
Calling snprintf with OS_FLSIZE + 1 while passing in a message of OS_FLSIZE length pr longer would cause the function to overwrite the byte after the buffer with '\0'. This is not happening because all of the calls are done with small strings, but I encourage this calls to be changed so that the given length passed into snprintf is OS_FLSIZE, improving code safety for the future.

Molter73 commented on src/config/active-response.c:
```diff
@@ -428,10 +593,206 @@ int ReadActiveCommands(XML_NODE node, void *d1, __attribute__((unused)) void *d2
+
+    XML_NODE node = OS_GetElementsbyNode(&xml, NULL);
+    XML_NODE child = NULL;
+    char *repeated_t = NULL;
```
*_t usually refer to typedefs, I encourage strongly for repeated_t to be renamed into repeated os something more meaningful.

Molter73 replied:
Also, repeated_a (declared below) should be renamed into something more meaningful to improve code readability.

SalvaMilenario commented:
Hi 

While testing the validator, I found this error. It occurs when it add this configuration block. 
``` XML 
  <email_alerts>
    <email_to>recipient@example.wazuh.com</email_to>
    <email_to>recipient2@example.wazuh.com</email_to>
    <level>12</level>
    <group>sshd,</group>
    <do_not_delay/>
    <email_lo_source>alerts.log</email_lo_source>
  </email_alerts>
```
This is the error: 
```
[root@centos7-manager1 vagrant]# ossec-control restart
2019/10/29 10:18:42 ossec-maild: ERROR: (1230): Invalid element in the configuration: 'email_lo_source'.
2019/10/29 10:18:42 ossec-maild: ERROR: (1202): Configuration error at '/var/ossec/etc/ossec.conf'.
2019/10/29 10:18:42 ossec-maild: CRITICAL: (1202): Configuration error at '/var/ossec/etc/ossec.conf'.
ossec-maild: Configuration error. Exiting
```
This error also happen in wazuh version 3.11. 

Regards

carlocas commented:
Hi @SalvaMilenario 

I have been checking the code and I found that the error that you reported is provoked by a missing option in the file `src/config/email-alerts-config.c`. 
https://github.com/wazuh/wazuh/blob/86ccb6cd55511e5d9727c8a8bfcd0465720e6371/src/config/email-alerts-config.c#L22-L31
How can you see in that file there isn't an option to `email_log_source`, this causes an error at the time to read the configuration. To solve this problem I recommended to put the missing option in the file and the correspondent comprobations to charge in memory.

Regards

SalvaMilenario commented:
Hi

While testing the validator, if respite an option in any block of the configuration it generated several memory leaks. In addition, it notified an error in the database that should not appear. This is the file `ossec.conf` what I used to generate the memory leaks:
``` XML
<ossec_config>
  <global>
    <jsonout_output>yes</jsonout_output>
    <alerts_log>yes</alerts_log>
    <logall>no</logall>
    <logall_json>no</logall_json>
    <email_notification>no</email_notification>
    <smtp_server>smtp.example.wazuh.com</smtp_server>
    <email_from>ossecm@example.wazuh.com</email_from>
    <email_to>recipient@example.wazuh.com</email_to>
    <email_maxperhour>12</email_maxperhour>
    <smtp_server>smtp.example.wazuh.com</smtp_server>>
  </global>

  <alerts>
    <log_alert_level>3</log_alert_level>
    <email_alert_level>12</email_alert_level>
  </alerts>

  <logging>
    <log_format>plain</log_format>
  </logging>

  <remote>
    <connection>secure</connection>
    <port>1514</port>
    <protocol>udp</protocol>
    <queue_size>131072</queue_size>
  </remote>

  <rootcheck>
    <disabled>no</disabled>
    <check_files>yes</check_files>
    <check_trojans>yes</check_trojans>
    <check_dev>yes</check_dev>
    <check_sys>yes</check_sys>
    <check_pids>yes</check_pids>
    <check_ports>yes</check_ports>
    <check_if>yes</check_if>

    <frequency>43200</frequency>

    <rootkit_files>/var/ossec/etc/rootcheck/rootkit_files.txt</rootkit_files>
    <rootkit_trojans>/var/ossec/etc/rootcheck/rootkit_trojans.txt</rootkit_trojans>

    <rootkit_files>/var/ossec/etc/rootcheck/rootkit_files.txt</rootkit_files>
    <rootkit_trojans>/var/ossec/etc/rootcheck/rootkit_trojans.txt</rootkit_trojans>

    <skip_nfs>yes</skip_nfs>
  </rootcheck>

  <wodle name="cis-cat">
    <disabled>yes</disabled>
    <timeout>1800</timeout>
    <interval>1d</interval>
    <scan-on-start>yes</scan-on-start>

    <java_path>wodles/java</java_path>
    <ciscat_path>wodles/ciscat</ciscat_path>

    <java_path>wodles/java</java_path>
    <ciscat_path>wodles/ciscat</ciscat_path>
  </wodle>

  <wodle name="osquery">
    <disabled>yes</disabled>
    <run_daemon>yes</run_daemon>
    <log_path>/var/log/osquery/osqueryd.results.log</log_path>
    <config_path>/etc/osquery/osquery.conf</config_path>
    <add_labels>yes</add_labels>
    <log_path>/var/log/osquery/osqueryd.results.log</log_path>
    <config_path>/etc/osquery/osquery.conf</config_path>
  </wodle>

  <syscheck>
    <disabled>no</disabled>

    <frequency>43200</frequency>

    <scan_on_start>yes</scan_on_start>

    <alert_new_files>yes</alert_new_files>

    <auto_ignore frequency="10" timeframe="3600">no</auto_ignore>

    <directories check_all="yes">/etc,/usr/bin,/usr/sbin</directories>
    <directories check_all="yes">/bin,/sbin,/boot</directories>

    <ignore>/etc/mtab</ignore>
    <ignore>/etc/hosts.deny</ignore>
    <ignore>/etc/mail/statistics</ignore>
    <ignore>/etc/random-seed</ignore>
    <ignore>/etc/random.seed</ignore>
    <ignore>/etc/adjtime</ignore>
    <ignore>/etc/httpd/logs</ignore>
    <ignore>/etc/utmpx</ignore>
    <ignore>/etc/wtmpx</ignore>
    <ignore>/etc/cups/certs</ignore>
    <ignore>/etc/dumpdates</ignore>
    <ignore>/etc/svc/volatile</ignore>
    <ignore>/sys/kernel/security</ignore>
    <ignore>/sys/kernel/debug</ignore>
    <ignore>/dev/core</ignore>

    <ignore type="sregex">^/proc</ignore>
    <ignore type="sregex">.log$|.swp$</ignore>

    <nodiff>/etc/ssl/private.key</nodiff>
    <nodiff>/etc/ssl/private.key</nodiff>

    <skip_nfs>yes</skip_nfs>
  </syscheck>

  <global>
    <white_list>127.0.0.1</white_list>
    <white_list>^localhost.localdomain$</white_list>
    <white_list>10.0.2.3</white_list>
  </global>

  <command>
    <name>route-null</name>
    <executable>route-null.sh</executable>
    <expect>srcip</expect>
    <timeout_allowed>yes</timeout_allowed>
    <name>route-null</name>
    <executable>route-null.sh</executable>
  </command>

  <localfile>
    <log_format>audit</log_format>
    <location>/var/log/audit/audit.log</location>
    <log_format>audit</log_format>
    <location>/var/log/audit/audit.log</location>
  </localfile>

  <localfile>
    <log_format>full_command</log_format>
    <command>netstat -tulpn | sed 's/\([[:alnum:]]\+\)\ \+[[:digit:]]\+\ \+[[:digit:]]\+\ \+\(.*\):\([[:digit:]]*\)\ \+\([0-9\.\:\*]\+\).\+\ \([[:digit:]]*\/[[:alnum:]\-]*\).*/\1 \2 == \3 == \4 \5/' | sort -k 4 -g | sed 's/ == \(.*\) ==/:\1/' | sed 1,2d</command>
    <command>netstat -tulpn | sed 's/\([[:alnum:]]\+\)\ \+[[:digit:]]\+\ \+[[:digit:]]\+\ \+\(.*\):\([[:digit:]]*\)\ \+\([0-9\.\:\*]\+\).\+\ \([[:digit:]]*\/[[:alnum:]\-]*\).*/\1 \2 == \3 == \4 \5/' | sort -k 4 -g | sed 's/ == \(.*\) ==/:\1/' | sed 1,2d</command>
    <alias>netstat listening ports</alias>
    <alias>netstat listening ports</alias>
    <frequency>360</frequency>
  </localfile>

  <!-- Configuration for ossec-authd -->
  <auth>
    <disabled>no</disabled>
    <port>1515</port>
    <use_source_ip>yes</use_source_ip>
    <force_insert>yes</force_insert>
    <force_time>0</force_time>
    <purge>yes</purge>
    <use_password>no</use_password>
    <limit_maxagents>yes</limit_maxagents>
    <ciphers>HIGH:!ADH:!EXP:!MD5:!RC4:!3DES:!CAMELLIA:@STRENGTH</ciphers>
    <!-- <ssl_agent_ca></ssl_agent_ca> -->
    <ssl_verify_host>no</ssl_verify_host>
    <ssl_manager_cert>/var/ossec/etc/sslmanager.cert</ssl_manager_cert>
    <ssl_manager_key>/var/ossec/etc/sslmanager.key</ssl_manager_key>
    <ssl_auto_negotiate>no</ssl_auto_negotiate>
    <ssl_manager_cert>/var/ossec/etc/sslmanager.cert</ssl_manager_cert>
    <ssl_manager_key>/var/ossec/etc/sslmanager.key</ssl_manager_key>
  </auth>

  <cluster>
    <name>wazuh</name>
    <node_name>node01</node_name>
    <node_type>master</node_type>
    <key></key>
    <port>1516</port>
    <bind_addr>0.0.0.0</bind_addr>
    <nodes>
        <node>NODE_IP</node>
    </nodes>
    <hidden>no</hidden>
    <disabled>yes</disabled>
    <name>wazuh</name>
    <node_name>node01</node_name>
    <node_type>master</node_type>

  </cluster>

  <agentless>
    <type>ssh_integrity_check_linux</type>
    <frequency>300</frequency>
    <host>admin@192.168.1.108</host>
    <state>periodic_diff</state>
    <arguments>/etc /usr/bin /usr/sbin</arguments>
  </agentless>

  <database_output>
    <hostname>192.168.1.122</hostname>
    <username>MySQLadmin</username>
    <password>secret1234</password>
    <database>Alerts_DB</database>
    <type>mysql</type>
    <hostname>192.168.1.122</hostname>
    <username>MySQLadmin</username>
    <password>secret1234</password>
    <database>Alerts_DB</database>
  </database_output>

  <integration>
    <name>slack</name>
    <hook_url>https://hooks.slack.com/services/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</hook_url>
    <level>10</level>
    <group>multiple_drops|authentication_failures</group>
    <alert_format>json</alert_format>
    <name>slack</name>
    <hook_url>https://hooks.slack.com/services/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</hook_url>
    <group>multiple_drops|authentication_failures</group>
    <alert_format>json</alert_format>
  </integration>

  <fluent-forward>
    <enabled>yes</enabled>
    <socket_path>/var/run/fluent.sock</socket_path>
    <address>localhost</address>
    <port>24224</port>
    <shared_key>secret_string</shared_key>
    <ca_file>/root/certs/fluent.crt</ca_file>
    <user>foo</user>
    <password>bar</password>
    <socket_path>/var/run/fluent.sock</socket_path>
    <address>localhost</address>
    <shared_key>secret_string</shared_key>
    <ca_file>/root/certs/fluent.crt</ca_file>
    <user>foo</user>
    <password>bar</password>
  </fluent-forward>

</ossec_config>
```
This is the log of Valgrind, where are all memory leaks:  
```
==5903== Memcheck, a memory error detector
==5903== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==5903== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==5903== Command: /var/ossec/bin/check_configuration -t manager -f /home/vagrant/provision/ossec.conf
==5903== 

ERROR: Invalid configuration in Database

==5903== 
==5903== HEAP SUMMARY:
==5903==     in use at exit: 828 bytes in 35 blocks
==5903==   total heap usage: 35,218 allocs, 35,183 frees, 8,982,436 bytes allocated
[...]
==5903== LEAK SUMMARY:
==5903==    definitely lost: 828 bytes in 35 blocks
==5903==    indirectly lost: 0 bytes in 0 blocks
==5903==      possibly lost: 0 bytes in 0 blocks
==5903==    still reachable: 0 bytes in 0 blocks
==5903==         suppressed: 0 bytes in 0 blocks
==5903== 
==5903== For lists of detected and suppressed errors, rerun with: -s
==5903== ERROR SUMMARY: 29 errors from 29 contexts (suppressed: 0 from 0)
``` 
Regards.

Molter73 commented on src/config/agentlessd-config.c:
```diff
@@ -146,15 +202,66 @@ int Read_CAgentless(XML_NODE node, void *config, __attribute__((unused)) void *c
+    return 0;
+}
+
+void free_AgentlessConfig(agentlessd_config *config) {
```
The whila loops in this function could be replaced by for loops, improving readability.

Molter73 commented on src/config/integrator-config.c:
```diff
@@ -158,40 +186,112 @@ int Read_Integrator(XML_NODE node, void *config, __attribute__((unused)) void *c
 }
+
+int Test_Integratord(const char *path, char **output) {
+    IntegratorConfig **test_integrator = NULL;
```
Should not test_integrator be an IntegratorConfig* instead of an IntegratorConfig**?

carlocas replied:
It should be an integratorConfig**, because if a user introduces more than one integrator module in ossec.conf, the program will save the data in a vector. So for free the pointer later will need that be of type Integrator Config**.

Molter73 commented on src/config/syscheck-config.c:
```diff
@@ -724,20 +940,33 @@ static int read_attr(syscheck_config *syscheck, const char *dirs, char **g_attrs
+                        "Glob error. Invalid pattern: '%s'.",
+                        expandedpath);
+                    wm_strcat(output, message, '\n');
+                }                ret = 1;
```
Could we move this ret = 1 to the following line?

Molter73 commented on src/config/active-response.c:
```diff
@@ -273,14 +383,23 @@ int ReadActiveResponses(XML_NODE node, void *d1, void *d2)
+            snprintf(message, OS_FLSIZE,
+                "Could not acquire memory due to [(%d)-(%s)].",
+                errno, strerror(errno));
+            wm_strcat(output, message, '\n');
```
The original code exited on this error, the new code keeps running, is this the intended behaviour?

Molter73 commented on src/config/wmodules-sca.c:
```diff
@@ -89,27 +101,55 @@ int wm_sca_read(const OS_XML *xml,xml_node **nodes, wmodule *module)
+                        nodes[i]->element, nodes[i]->content);
+                    wm_strcat(output, message, '\n');
+                }
+                return (OS_INVALID);
```
Fix indentation in this block of code please.

Molter73 commented on src/config/wmodules-sca.c:
```diff
@@ -189,11 +257,25 @@ int wm_sca_read(const OS_XML *xml,xml_node **nodes, wmodule *module)
+                            merror("Empty policy value at '%s'.", WM_SCA_CONTEXT.name);
+                        } else {
+                            snprintf(message, OS_FLSIZE,
+                                "Policy path is too long at module '%s'. Max path length is %d",
```
This message is different from the one on the merror call.

Molter73 commented on src/shared/check_config.c:
```diff
@@ -0,0 +1,171 @@
+
+    /* Read the main configuration */
+    i = 0;
+    while (node[i]) {
```
We could replace this while loop with a for loop for better readability

Molter73 commented on src/wazuh_modules/main.c:
```diff
@@ -1,4 +1,4 @@
-/*
+    /*
```
Please remove this 4 spaces

Molter73 commented:
https://github.com/wazuh/wazuh/blob/7af46021e51a2c4721ef2a0ce5bc5bd16b1e6769/src/wazuh_modules/wm_check_config.c#L118-L147
This while loop could be refactored into a for loop. Also, aux could be renamed into something more expressive like line or current_line and aux_2 can be removed if the calls to cJSON_AddStringToObject(validator, "message", aux_2) are replaced by cJSON_AddStringToObject(validator, "message", strdup(aux + offset)) where offset is the length of the found tag, which would also avoid unnecesary moves inside the duplicated string.

carlocas commented:
Hi team

The function `wm_destroy` only free the pointer of the variable and not the information inside it. For that reason, I added several free in the code.

https://github.com/wazuh/wazuh/blob/86ccb6cd55511e5d9727c8a8bfcd0465720e6371/src/wazuh_modules/wm_fluent.c#L134-L138

Also, the remaining memory leaks detected by @SalvaMilenario were caused by overwriting the variables without been freed previously. So I introduced a condition to check if the variable is not NULL. It the condition is true, free the memory and add the new valor. 

Regards.

SalvaMilenario commented:
Hi @carlocas 

I'm still getting this memory leak.
``` XML
  <agentless>
    <type>ssh_integrity_check_linux</type>
    <frequency>300</frequency>
    <host>admin@192.168.1.108</host>
    <state>periodic_diff</state>
    <arguments>/etc /usr/bin /usr/sbin</arguments>
    <type>ssh_integrity_check_linux</type>
    <frequency>300</frequency>
    <host>admin@192.168.1.108</host>
    <state>periodic_diff</state>
    <arguments>/etc /usr/bin /usr/sbin</arguments>
  </agentless>
```
The Valgrind log: 
```
==14693== Memcheck, a memory error detector
==14693== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==14693== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==14693== Command: /var/ossec/bin/check_configuration -t manager -f /home/vagrant/provision/ossec.conf
==14693== 

Configuration validated successfully

==14693== 
==14693== HEAP SUMMARY:
==14693==     in use at exit: 26 bytes in 1 blocks
==14693==   total heap usage: 38,364 allocs, 38,363 frees, 10,046,149 bytes allocated
==14693== 
==14693== 26 bytes in 1 blocks are definitely lost in loss record 1 of 1
==14693==    at 0x4C29EA3: malloc (vg_replace_malloc.c:309)
==14693==    by 0x5CF0AA9: strdup (strdup.c:42)
==14693==    by 0x4382CC: Read_CAgentless (agentlessd-config.c:165)
==14693==    by 0x42134D: read_main_elements (config.c:98)
==14693==    by 0x422154: ReadConfig (config.c:274)
==14693==    by 0x4389F9: Test_Agentlessd (agentlessd-config.c:232)
==14693==    by 0x406BF9: test_manager_conf (check_config.c:42)
==14693==    by 0x40689D: main (check_configuration.c:116)
==14693== 
==14693== LEAK SUMMARY:
==14693==    definitely lost: 26 bytes in 1 blocks
==14693==    indirectly lost: 0 bytes in 0 blocks
==14693==      possibly lost: 0 bytes in 0 blocks
==14693==    still reachable: 0 bytes in 0 blocks
==14693==         suppressed: 0 bytes in 0 blocks
==14693== 
==14693== For lists of detected and suppressed errors, rerun with: -s
==14693== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```
Regards

Molter73 commented on src/.gitignore:
```diff
@@ -0,0 +1,2 @@
+check_configuration
+src/check_configuration
```
Please move the content of this file into the "Compile Programs" section of wazuh/.gitignore and remove this file.

SalvaMilenario commented on src/config/remote-config.c:
```diff
@@ -238,9 +354,87 @@ int Read_Remote(XML_NODE node, void *d1, __attribute__((unused)) void *d2)
+
+        int i = 0;
+        if(rmt->allowips) {
+            while(rmt->allowips[i]) {
```
This loop `while` behaves as a `for` change it, please.

SalvaMilenario commented on src/shared/list_op.c:
```diff
@@ -334,3 +334,12 @@ void *OSList_AddData(OSList *list, void *data)
+
+void Free_OSList(OSList *list) {
+    OSListNode *node = OSList_GetFirstNode(list);
+    while(node) {
```
It`d more readable if you will use the loop `for`.

SalvaMilenario commented on src/wazuh_modules/wm_sca.c:
```diff
@@ -1830,7 +1830,17 @@ static int wm_sca_is_process(char *value, OSList *p_list)
+            os_free(data->profile[i]->policy_id);
+            os_free(data->profile[i]);
+        }
+        os_free(data->profile);
```
This `os_free` check if `data->profile` is NULL, but in the line 1836 yet has been checked this. I recommend what you used `free` for not check again.

Molter73 commented:
During testing of the check_configuration module to verify the email_alerts tag with the attached configuration, the module runs infinitely.
```
  <email_alerts>
    <level>12</level>
  </email_alerts>
```
The problem seems to be that not having a <email_to> tag causes granto_email_counter to be initialized to 0, causing the following condition to never be met
https://github.com/wazuh/wazuh/blob/3d231ffd3d785ea9604af6e694d006c6147157f1/src/config/email-alerts-config.c#L252

carlocas commented:
Hi team

In the last commits, I have been working in improving the readability of the code, changing multiple while loop, that behaves like a for, for a loop for. Also, I was working in the bug found by @Molter73, this bug was provoked by a variable defined as `unsigned int`. With the tag `email_to` in `email_alerts`, we don't any problem. Everything work as expected, but if we don't put this tag. The program initializes the variable to zero, and then deduct one. These results provoked an overflow, so instead of decrement the variable, we obtained the maximum numeric valor for that variable and enter in an infinite loop. The result is that the program allocating memory until freeze the computer.  To fix this problem I changed the type of the variable of unsigned int to int.

Related commit:
[4c23dc3](https://github.com/wazuh/wazuh/pull/4139/commits/4c23dc3582e208aef035c61a4bb61512c341fba3)

Regards.

Molter73 commented:
I've been running check_configuration against a ossec.conf file, adding different parameters and trying them out according to the documentation and found some behaviour that either seems wrong or is not congruent with the documentation. (Please keep in mind that I didn't check if the configurations actually worked, I simply checked what the output of check_configuration is).

### manager
- global.email_to: When granular configuration is used, the global.email_to is not checked for existance
- global.email_from: does not check for a valid email.
- global.email_reply_to: does not check for a valid email.
- global.smtp_server: content is not checked.
- global.helo_server: content is not checked.
- global.email_maxperhour: when this configuration is wrong, the following message is triggered: WARNING: Detected a deprecated configuration for cluster. Interval option is not longer available.
- global.email_idsname: content is not checked.
- global.email_log_source: only checks for an empty tag and numbers, not alerts.log | alerts.json
- global.custom_alert_output: content is not checked.
- global.stats: does not check for empty tags.
- global.white_list: content is not checked.
- global.host_information: does not check for null and has no upper bound limit (16 is specified in the docs).
- global.prelude_log_level: does not check for null and has no upper bound limit (16 is specified in the docs).
- global.zeromq_uri: content is not checked.
- global.geoipdb: content is not checked.
- global.rotate_interval: the minimum interval is 10m, not 10s as specified by the docs.
- alerts.log_alert_level: allows any positive number, 1-16 is specified in the docs.
- alerts.email_alert_level: allows any positive number, 1-16 is specified in the docs.
- email_alerts.group: content is not checked.
- email_alerts.event_location: content is not checked.
- email_alerts.rule_id: content is not checked.
- email_alerts.do_not_delay: content is not checked, must be null.
- email_alerts.do_not_group: content is not checked, must be null.
- email_alerts.email_log_source: could not make this rule work.
- command.expect: content is not checked.
- command.extra_args: content is not checked.
- auth.port: no restriction on port number.
- auth.force_time: allows negative numbers.
- auth.ssl_agent_ca: content is not checked.
- auth.ciphers: content is not checked.
- auth.ssl_manager_cert: content is not checked.
- cluster.key: content is not checked.
- cluster.interval: content is not checked.
- cluster.port: content is not checked.
- cluster.bind_addr: content is not checked.
- cluster.nodes: if this tag is empty, triggers: WARNING: Detected a deprecated configuration for cluster. Interval option is not longer available.
- cluster.nodes.node: content is not checked.
- agentless.type: this tag is allowed to be empty.
- agentless.host: content is not checked.

Not all of the remarks are necesarily wrong, but seems like a good idea to go over them and check if the behaviours are ok. I will continue to test other configurations and update this comment accordingly.

Molter73 commented:
Running check_configuration with the following ossec.config configuration causes a core dump.
```
  <labels>
    <label></label>
  </labels>
```
The problem seems to be that labels_free is being called twice and attempting to free a pointer that was already released.
It is first called at 
https://github.com/wazuh/wazuh/blob/29d45fbf66f2f32388f66bd6c6654fd56cdad281/src/config/labels-config.c#L139-L142
and later at 
https://github.com/wazuh/wazuh/blob/29d45fbf66f2f32388f66bd6c6654fd56cdad281/src/config/global-config.c#L1063

Molter73 commented:
Running check_configuration with the following configuration causes a segmentation fault
```
  <localfile>
    <log_format>audit</log_format>
  </localfile>
```
```
[root@manager bin]# ./check_configuration -t manager -f /home/vagrant/ossec.conf

Segmentation fault (core dumped)
```
The problem seems to be in this lines of code
https://github.com/wazuh/wazuh/blob/29d45fbf66f2f32388f66bd6c6654fd56cdad281/src/config/localfile-config.c#L546-L550
logf[pl].file has not been initialized and calling strchr on it causes the segmentation fault.

Molter73 commented:
If the configuration checked does not have any errors, any warnings that were issued during the validation are not printed.
https://github.com/wazuh/wazuh/blob/29d45fbf66f2f32388f66bd6c6654fd56cdad281/src/util/check_configuration.c#L124-L128
I recommend we change this code, so that it prints output regardless of the result and print the success message on result == 0. That way a successful configuration will be notified along with any warnings it has.

Molter73 commented on src/config/syscheck-config.c:
```diff
@@ -724,20 +940,34 @@ static int read_attr(syscheck_config *syscheck, const char *dirs, char **g_attrs
             if (g.gl_pathv[0] == NULL) {
-                merror(GLOB_NFOUND, expandedpath);
+                if (output == NULL){
+                    merror(GLOB_ERROR, expandedpath);
```
Was this changed from GLOB_NFOUND to GLOB_ERROR on purpouse?

SalvaMilenario commented on src/config/active-response.c:
```diff
@@ -54,35 +56,67 @@ int ReadActiveResponses(XML_NODE node, void *d1, void *d2)
+                DEFAULTARPATH, errno, strerror(errno));
+            wm_strcat(output, message, '\n');
+        }
         return (-1);
```
For unifying the values always use `return (OS_INVALID);`

SalvaMilenario commented on src/config/active-response.c:
```diff
@@ -428,10 +600,200 @@ int ReadActiveCommands(XML_NODE node, void *d1, __attribute__((unused)) void *d2
 
+    if (output) {
+        if (tmp_command->name) {
+            free (tmp_command->name);
```
Please, change it, and use `os_free` that already does that check and we avoid repeating code.

SalvaMilenario commented on src/config/agentlessd-config.c:
```diff
@@ -146,15 +205,63 @@ int Read_CAgentless(XML_NODE node, void *config, __attribute__((unused)) void *c
+
+void free_AgentlessConfig(agentlessd_config *config) {
+    if(config) {
+        int i, j;
```
The scope of the variables `i` and `j` should be inside the conditional `if(config->entries)`

SalvaMilenario commented on src/config/authd-config.c:
```diff
@@ -37,10 +37,17 @@ int Read_Authd(XML_NODE node, void *d1, __attribute__((unused)) void *d2) {
     snprintf(manager_cert, OS_SIZE_1024 - 1, "%s/etc/sslmanager.cert", DEFAULTDIR);
     snprintf(manager_key, OS_SIZE_1024 - 1, "%s/etc/sslmanager.key", DEFAULTDIR);
 
+    if (output) {
```
Remove this condition, it doesn't necessarily.

SalvaMilenario commented on src/config/authd-config.c:
```diff
@@ -187,3 +281,32 @@ short eval_bool(const char *str) {
+        } else {
+            wm_strcat(output, "ERROR: Invalid configuration in Authd", '\n');
+        }
+        free_authd_config(test_authd);
```
This sentence is the same as line 300. Please, modify it so that it doesn't repeat itself.

SalvaMilenario commented on src/config/buffer-config.c:
```diff
@@ -36,78 +37,145 @@ int Read_ClientBuffer(XML_NODE node, __attribute__((unused)) void *d1, void *d2)
+        } else{
+            wm_strcat(output, "ERROR: Invalid configuration in ClientBuffer", '\n');
+        }
+        Free_Client(&test_clientBuffer);
```
This sentence is the same as line 179.  Please, modify it so that it doesn't repeat itself.

SalvaMilenario commented on src/config/client-config.c:
```diff
@@ -49,94 +50,200 @@ int Read_Client(const OS_XML *xml, XML_NODE node, void *d1, __attribute__((unuse
         rip = NULL;
```
This variable is now null. It's not necessary to reassign  `NULL`.

carlocas replied:
This variable is assigned to `NULL` at the beginning of a loop for. It can be changed inside that loop, so it is recommendable to reassign to `NULL`.

SalvaMilenario commented on src/config/cluster-config.c:
```diff
@@ -87,18 +165,67 @@ int Read_Cluster(XML_NODE node, void *d1, __attribute__((unused)) void *d2) {
+                /* Get children */
+                xml_node **children = NULL;
+                    if (children = OS_GetElementsbyNode(xml, node[i]), !children) {
+                return OS_INVALID;
```
Before a `return OS_INVALID`, should be an error message. Please, notify the error.

SalvaMilenario commented on src/config/config.c:
```diff
@@ -378,3 +476,54 @@ int ReadConfig(int modules, const char *cfgfile, void *d1, void *d2)
+            } else {
+                wm_strcat(output, "ERROR: Invalid configuration in Analysisd", '\n');
+            }
+            fail = 1;
```
This variable is now 1. It's not necessary to reassign 1.

SalvaMilenario commented on src/config/config.c:
```diff
@@ -378,3 +476,54 @@ int ReadConfig(int modules, const char *cfgfile, void *d1, void *d2)
+            } else {
+                wm_strcat(output, "Maximum output size must be between 1 MiB and 1 TiB.", '\n');
+            }
+            fail = 1;
```
This variable is now 1. It's not necessary to reassign 1.

SalvaMilenario commented on src/config/config.c:
```diff
@@ -378,3 +476,54 @@ int ReadConfig(int modules, const char *cfgfile, void *d1, void *d2)
+            fail = 1;
+        }
+
+        if (!fail && test_config->max_output_size && (test_config->max_output_size < 1000000 || test_config->max_output_size > 1099511627776)) {
```
The condition `!fail` has already been evaluated previously. Please, removed it.

SalvaMilenario commented on src/config/mail-config.c:
```diff
@@ -0,0 +1,132 @@
+fail:
+    config_free(global_config);
+    freeMailConfig(mail_config);
+    if (output) {
```
Remove this condition, it isn't necessary.

SalvaMilenario commented on src/config/rootcheck-config.c:
```diff
@@ -191,105 +274,187 @@ int Read_Rootcheck(XML_NODE node, void *configp, __attribute__((unused)) void *m
+        } else {
+            wm_strcat(output, "ERROR: Invalid configuration in Rootcheck", '\n');
+        }
+	    Free_Rootcheck(&test_rootcheck);
```
This sentence is the same as line 453. Please, modify it so that it doesn't repeat itself.

SalvaMilenario commented on src/config/rules-config.c:
```diff
@@ -113,14 +114,44 @@ int Read_Rules(XML_NODE node, void *configp, __attribute__((unused)) void *mailp
     regex.mutex = (pthread_mutex_t) PTHREAD_MUTEX_INITIALIZER;
 
+    if (output){
+        int i;
```
This variable is now declared. Removed it.

SalvaMilenario commented on src/config/rules-config.c:
```diff
@@ -113,14 +114,44 @@ int Read_Rules(XML_NODE node, void *configp, __attribute__((unused)) void *mailp
+    }
+
     if (node) {
         while (node[i]) {
```
The value of `i` must be restarted for the correct loop operation.

SalvaMilenario commented on src/wazuh_modules/wm_ciscat.c:
```diff
@@ -1625,6 +1625,14 @@ void wm_ciscat_destroy(wm_ciscat *ciscat) {
         free(cur_eval);
     }
 
+    if (ciscat->ciscat_path){
```
Please, change it, and use os_free that already does that check and we avoid repeating code.

SalvaMilenario commented on src/wazuh_modules/wm_sca.c:
```diff
@@ -1830,7 +1830,17 @@ static int wm_sca_is_process(char *value, OSList *p_list)
-void wm_sca_destroy(wm_sca_t * data) {
+void wm_sca_destroy(wm_sca_t *data) {
+
+    int i;
```
The scope of the variable `i` should be inside the conditional if(data->profile)

Molter73 commented:
Running check_configuration with the following fluent-forward configuration causes a Segmentation fault
```
<fluent-forward>
  <enabled>yes</enabled>
</fluent-forward>
```
This is the line that causes the segmentation fault.
https://github.com/wazuh/wazuh/blob/1e44c66a118768312e5747bad3993991f9ab4b1a/src/wazuh_modules/wm_fluent.c#L137
When no address is specified, it is initialized with a static string, the easiest solution would be to change this initialization to a strdup of the same string.
https://github.com/wazuh/wazuh/blob/1e44c66a118768312e5747bad3993991f9ab4b1a/src/config/wmodules-fluent.c#L340-L343

Molter73 commented:
https://github.com/wazuh/wazuh/blob/2a01962eb53a29d1da9e6af67049bff45783330e/src/config/wmodules-oscap.c#L249-L262
When a wrong configuration for oval-id is given, this code reports and error on the xccdf-id tag

Molter73 commented:
Placing a wodle pair of tags with no name attribute causes a segmentation fault `<wodle></wodle>`
The problem is caused by trying the following line, attribute is dereferenced while being null.
https://github.com/wazuh/wazuh/blob/2a01962eb53a29d1da9e6af67049bff45783330e/src/config/wmodules-config.c#L27

Molter73 commented:
While testing vulnerability-detector after the rebase to version 3.11 I found that leaving the allow tag empty on either an nvd or redhat provider causes a segmentation fault on the following line:
https://github.com/wazuh/wazuh/blob/abdea9bd59ee424b44aa31b1d231c2113961fdb9/src/config/wmodules-vuln-detector.c#L1014
This is caused due to `*node[i]->values` not being initialized before hand.

Molter73 commented on src/config/config.c:
```diff
@@ -374,7 +465,14 @@ int ReadConfig(int modules, const char *cfgfile, void *d1, void *d2)
+                merror(XML_INVELEM, node[i]->element);
+            } else {
+                snprintf(message, OS_FLSIZE,
+                    "Invalid elemente in the configuration: '%s'.",
```
Typo: elemente

carlocas commented:
Hi team:

In the last commits, we have been working to fix some segfaults that appeared when introduced the next wrong configurations:
```
<labels>
  <labels></labels>
</labels>

<localfile>
  <log_format>audit</log_fortmat>
</localfile>

<fluent-forward>
  <enabled>yes</enabled>
</fluent-forward>

<wodle></wodle>
```
Also, the wrong configuration of `<allow></allow>` in vulnerability detector, has been fixed in the branch of vulnerability detector for wazuh 3.11.

Regards
